### ISSUE TITLE
Improve NATS envelope header

### ISSUE THREAD

**Issue by llchan** [liftbridge-io/liftbridge#128]
- The current envelope cookie `"LIFT"` has an above-average likelihood of appearing at the beginning of a nats message, since it's ascii (and an English word). I think the magic number should ideally be 4 random bytes that are _not_ valid UTF-8.
- Should we add a checksum to further reduce the probability of an accidental lift message?  I think a CRC32 of the payload is probably sufficient.  Flatbuffers-wise it's actually more convenient to put this in the header rather than footer, but that's probably something I could hack around if we want it in the footer.

**Comment by llchan** [comment]
Just pulled this off urandom, and confirmed it's not valid utf-8 and not in the wikipedia [list of file signatures](https://en.wikipedia.org/wiki/List_of_file_signatures).
```
b9 0e 43 b4
```

I'd recommend following that with a 1-byte protocol version, so that we have the ability to update the format later in a safe way.  1 byte should be sufficient because this would really only be for envelope header changes or major non-backwards-compatible proto/flatbuffers changes (most of the normal stuff like adding fields will be handled by the encoding).

Along with that, typically I've seen people use another byte to specify the header length. I think sparing 1 byte is reasonable for safety here.

Following that can be a 1-byte spot for bit flags (e.g. does the header contain a CRC32). And finally one more reserved byte to pad it up to a word.

To summarize, maybe this can be a proposed message header:
```
0               8               16              24              32
+---------------+---------------+---------------+---------------+
|       B9      |       0E      |       43      |       B4      |
+---------------+---------------+---------------+---------------+
|    Version    |   HeaderLen   |     Flags     |    Reserved   |
+---------------+---------------+---------------+---------------+
|                        CRC32 (optional)                       |
+---------------------------------------------------------------+

```
16 bytes by default, and if the user chooses to CRC32 it's a 24-byte header.

**Comment by llchan** [comment]
Actually maybe we should use that reserved byte for the message type, so we can allow the possibility of different lift messages without making a wrapper proto/flatbuffer to union all the possible "top level" message types.

**Comment by llchan** [comment]
Also since the checksum field doesnt exist yet, I might suggest using CRC-32C (aka Castagnoli, vs the IEEE one I see used elsewhere in liftbridge). That one has a dedicated SSE4.2 instruction that improves throughput.  The checksum probably isn't the bottleneck, but since it's a protocol-level decision we may as well choose the more hardware-friendly one up front.

**Comment by tylertreat** [comment]
The proposed header protocol makes sense to me, as does the checksumming. Since the work you're doing in #87 is going to have breaking changes, does it make sense to roll this change into that?

**Comment by llchan** [comment]
Yeah I think it makes sense to do all the protocol-level changes in one PR.  I'll try to get it pushed up as soon as I can for visibility.

**Comment by llchan** [comment]
It might also make sense to apply this header to replication messages for consistency.  To generalize a bit, maybe we could frame all liftbridge-related messages sent over nats with this header?  In the (hopefully rare) event that two different versions of liftbridge collide on one nats deployment/namespace, the version byte in the header would at least allow them to ignore messages from the other version and prevent possible decoding errors (or worse, data corruption).

**Comment by tylertreat** [comment]
Yeah, I was intending to re-work the replication protocol. I especially don't like it now because there is no version header in the current protocol. I think marrying the two to use the same protocol is a good idea.

**Comment by tylertreat** [comment]
This would also (I think) allow replication reads and client reads to use roughly the same code path. Replication currently just ships the bytes as is while client reads use protobuf/grpc. If those could work the same, or at least more similarly, that would be great.

**Comment by tylertreat** [comment]
This has largely been addressed in https://github.com/liftbridge-io/liftbridge/pull/153. Keeping the issue open for now though as there are still plans to change replication to use the new protocol.

**Comment by tylertreat** [comment]
Fully addressed in #154.
